snuba/datasets/entity_subscriptions/validators.py

```diff
@@ -0,0 +1,58 @@
+from abc import abstractmethod
+from typing import Any, Mapping, Sequence, Type, Union, cast
+
+from snuba.query.composite import CompositeQuery
+from snuba.query.data_source.simple import Entity
+from snuba.query.logical import Query
+from snuba.query.validation.validators import (
+    NoTimeBasedConditionValidator,
+    SubscriptionAllowedClausesValidator,
+)
+from snuba.utils.registered_class import RegisteredClass
+
+
+class InvalidSubscriptionError(Exception):
+    pass
+
+
+class EntitySubscriptionValidator(metaclass=RegisteredClass):
+    @classmethod
+    def config_key(cls) -> str:
+        return cls.__name__
+
+    @classmethod
+    def get_from_name(cls, name: str) -> Type["EntitySubscriptionValidator"]:
+        return cast(Type["EntitySubscriptionValidator"], cls.class_from_name(name))
+
+    @abstractmethod
+    def to_dict(self) -> Mapping[str, Any]:
+        raise NotImplementedError
+
+    @abstractmethod
+    def validate(self, query: Union[CompositeQuery[Entity], Query]) -> None:
+        raise NotImplementedError
+
+
+class AggregationValidator(EntitySubscriptionValidator):
+    def __init__(
+        self, max_allowed_aggregations: int, disallowed_aggregations: Sequence[str]
+    ):
+        self.max_allowed_aggregations = max_allowed_aggregations
+        self.disallowed_aggregations = disallowed_aggregations
+
+    def to_dict(self) -> Mapping[str, Any]:
+        return {}
+
+    def validate(self, query: Union[CompositeQuery[Entity], Query]) -> None:
+        from snuba.datasets.entities.factory import get_entity
```

Comment:
I assume this is avoiding some kind of circular import? What was the actual problem? Generally when this happens one of two things is wrong:

1) A circular import is being caused solely by typing (the easy case). We can use the `TYPE_CHECKING` constant in the imports to avoid that case.
2) Our code is organized poorly (the hard case). We need to move code around to fix this.

Reply:
Replied below.